File: claudesync/discovery.py
"""Scanner that discovers everything Claude Code stores on a machine.

Two scopes are scanned:

* **global** -- ``~/.claude/`` (or ``$CLAUDE_CONFIG_DIR``) plus the
  top-level ``~/.claude.json`` config file.
* **project** -- every directory Claude Code has ever been opened in.
  That list comes straight from ``~/.claude.json``'s ``"projects"`` map,
  which is the same source Claude Code Organizer uses, so results line up
  with what that tool shows.

Each discovered item is a :class:`DiscoveredItem`: either a single file
(memory, settings, an individual agent/command/rule/plan) or a whole
directory that should be copied as a unit (a skill, the plugin cache, the
session transcript store).
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, Iterator

GLOBAL_SCOPE = "global"


def claude_home(home: Path | None = None) -> Path:
    """Return Claude Code's global config directory.

    Honors ``$CLAUDE_CONFIG_DIR`` when set, otherwise ``<home>/.claude``.
    """
    override = os.environ.get("CLAUDE_CONFIG_DIR")
    if override:
        return Path(override).expanduser()
    return (home or Path.home()) / ".claude"


def claude_json_path(home: Path | None = None) -> Path:
    return (home or Path.home()) / ".claude.json"


@dataclass(frozen=True)
class DiscoveredItem:
    category: str  # memory | settings | mcp | agent | command | rule | plan | skill | session | plugin
    scope: str  # GLOBAL_SCOPE, or the absolute path of a project
    source: Path  # absolute path on disk (file or directory)
    is_dir: bool

    @property
    def label(self) -> str:
        return f"[{self.category}] {self.scope} -> {self.source}"


# (category, relative glob, is_dir_unit)
# is_dir_unit=True means the whole matched directory is copied as one item
# instead of being expanded file-by-file.
_GLOBAL_SPECS: tuple[tuple[str, str, bool], ...] = (
    ("memory", "CLAUDE.md", False),
    ("memory", "CLAUDE.local.md", False),
    ("settings", "settings.json", False),
    ("settings", "settings.local.json", False),
    ("mcp", "mcp.json", False),
    ("agent", "agents/**/*.md", False),
    ("command", "commands/**/*.md", False),
    ("rule", "rules/**/*.md", False),
    ("plan", "plans/**/*.md", False),
    ("skill", "skills/*", True),
    ("plugin", "plugins/*", True),
    ("session", "projects", True),
)

_PROJECT_SPECS: tuple[tuple[str, str, bool], ...] = (
    ("memory", "CLAUDE.md", False),
    ("memory", "CLAUDE.local.md", False),
    ("mcp", ".mcp.json", False),
    ("settings", ".claude/settings.json", False),
    ("settings", ".claude/settings.local.json", False),
    ("agent", ".claude/agents/**/*.md", False),
    ("command", ".claude/commands/**/*.md", False),
    ("rule", ".claude/rules/**/*.md", False),
    ("plan", ".claude/plans/**/*.md", False),
    ("skill", ".claude/skills/*", True),
)


def _expand(root: Path, category: str, pattern: str, is_dir_unit: bool) -> Iterator[Path]:
    if "*" not in pattern:
        candidate = root / pattern
        if candidate.exists():
            yield candidate
        return

    for match in sorted(root.glob(pattern)):
        if is_dir_unit and not match.is_dir():
            continue
        if not is_dir_unit and not match.is_file():
            continue
        yield match


def discover_projects(home: Path | None = None) -> list[Path]:
    """Return every project directory recorded in ``~/.claude.json``.

    Missing or unparsable config yields an empty list rather than raising,
    since a fresh machine may not have one yet.
    """
    path = claude_json_path(home)
    if not path.is_file():
        return []
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return []

    projects = data.get("projects")
    if not isinstance(projects, dict):
        return []

    result = []
    for raw_path in projects:
        candidate = Path(raw_path).expanduser()
        if candidate.is_dir():
            result.append(candidate)
    return result


def scan_global(home: Path | None = None) -> Iterator[DiscoveredItem]:
    home = home or Path.home()
    root = claude_home(home)

    claude_json = claude_json_path(home)
    if claude_json.is_file():
        yield DiscoveredItem("mcp", GLOBAL_SCOPE, claude_json, False)

    if not root.is_dir():
        return

    for category, pattern, is_dir_unit in _GLOBAL_SPECS:
        for match in _expand(root, category, pattern, is_dir_unit):
            yield DiscoveredItem(category, GLOBAL_SCOPE, match, match.is_dir())


def scan_project(project_path: Path) -> Iterator[DiscoveredItem]:
    project_path = Path(project_path)
    if not project_path.is_dir():
        return

    for category, pattern, is_dir_unit in _PROJECT_SPECS:
        for match in _expand(project_path, category, pattern, is_dir_unit):
            yield DiscoveredItem(category, str(project_path), match, match.is_dir())


def scan_all(
    extra_projects: Iterable[Path] = (),
    home: Path | None = None,
) -> list[DiscoveredItem]:
    """Scan the global scope plus every known (and extra) project."""
    items: list[DiscoveredItem] = list(scan_global(home))

    seen_projects: set[Path] = set()
    for project in list(discover_projects(home)) + list(extra_projects):
        project_path = Path(project).expanduser()
        # Resolve only to dedupe (so a symlinked or short-path alias of a
        # project already seen isn't scanned twice); keep the unresolved
        # path as the item's scope so it matches what's actually recorded
        # in ~/.claude.json instead of an OS-normalized alias of it.
        resolved = project_path.resolve()
        if resolved in seen_projects:
            continue
        seen_projects.add(resolved)
        items.extend(scan_project(project_path))

    return items

File: desktop/src/core/discovery.js
'use strict';

/**
 * Discovers everything Claude Code stores on this machine: the global
 * ~/.claude directory plus every project directory Claude Code has ever
 * been opened in (read from ~/.claude.json's "projects" map — the same
 * source Claude Code Organizer uses, and the same source the ClaudeSync
 * CLI uses, so results and backups line up between the two).
 */

const fs = require('fs');
const os = require('os');
const path = require('path');
const { glob } = require('./glob');

const GLOBAL_SCOPE = 'global';

function claudeHome(home) {
  if (process.env.CLAUDE_CONFIG_DIR) {
    return untildify(process.env.CLAUDE_CONFIG_DIR);
  }
  return path.join(home || os.homedir(), '.claude');
}

function untildify(p) {
  if (p.startsWith('~')) {
    return path.join(os.homedir(), p.slice(1));
  }
  return p;
}

function claudeJsonPath(home) {
  return path.join(home || os.homedir(), '.claude.json');
}

// [category, relative glob pattern, isDirUnit]
const GLOBAL_SPECS = [
  ['memory', 'CLAUDE.md', false],
  ['memory', 'CLAUDE.local.md', false],
  ['settings', 'settings.json', false],
  ['settings', 'settings.local.json', false],
  ['mcp', 'mcp.json', false],
  ['agent', 'agents/**/*.md', false],
  ['command', 'commands/**/*.md', false],
  ['rule', 'rules/**/*.md', false],
  ['plan', 'plans/**/*.md', false],
  ['skill', 'skills/*', true],
  ['plugin', 'plugins/*', true],
  ['session', 'projects', true],
];

const PROJECT_SPECS = [
  ['memory', 'CLAUDE.md', false],
  ['memory', 'CLAUDE.local.md', false],
  ['mcp', '.mcp.json', false],
  ['settings', '.claude/settings.json', false],
  ['settings', '.claude/settings.local.json', false],
  ['agent', '.claude/agents/**/*.md', false],
  ['command', '.claude/commands/**/*.md', false],
  ['rule', '.claude/rules/**/*.md', false],
  ['plan', '.claude/plans/**/*.md', false],
  ['skill', '.claude/skills/*', true],
];

function isDir(p) {
  try {
    return fs.statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function isFile(p) {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}

function expand(root, pattern, isDirUnit) {
  if (!pattern.includes('*')) {
    const candidate = path.join(root, pattern);
    return (isDirUnit ? isDir(candidate) : isFile(candidate)) ? [candidate] : [];
  }

  return glob(root, pattern).filter((match) => (isDirUnit ? isDir(match) : isFile(match)));
}

/** Every project directory recorded in ~/.claude.json, filtered to ones that still exist. */
function discoverProjects(home) {
  const jsonPath = claudeJsonPath(home);
  if (!isFile(jsonPath)) return [];

  let data;
  try {
    data = JSON.parse(fs.readFileSync(jsonPath, 'utf8'));
  } catch {
    return [];
  }

  const projects = data && typeof data.projects === 'object' ? data.projects : null;
  if (!projects) return [];

  return Object.keys(projects)
    .map((p) => untildify(p))
    .filter((p) => isDir(p));
}

function scanGlobal(home) {
  const items = [];
  const resolvedHome = home || os.homedir();

  const jsonPath = claudeJsonPath(resolvedHome);
  if (isFile(jsonPath)) {
    items.push({ category: 'mcp', scope: GLOBAL_SCOPE, source: jsonPath, isDir: false });
  }

  const root = claudeHome(resolvedHome);
  if (!isDir(root)) return items;

  for (const [category, pattern, isDirUnit] of GLOBAL_SPECS) {
    for (const match of expand(root, pattern, isDirUnit)) {
      items.push({ category, scope: GLOBAL_SCOPE, source: match, isDir: isDirUnit });
    }
  }

  return items;
}

function scanProject(projectPath) {
  const items = [];
  if (!isDir(projectPath)) return items;

  for (const [category, pattern, isDirUnit] of PROJECT_SPECS) {
    for (const match of expand(projectPath, pattern, isDirUnit)) {
      items.push({ category, scope: projectPath, source: match, isDir: isDirUnit });
    }
  }

  return items;
}

function scanAll({ extraProjects = [], home } = {}) {
  const resolvedHome = home || os.homedir();
  const items = scanGlobal(resolvedHome);

  const seen = new Set();
  const allProjects = [...discoverProjects(resolvedHome), ...extraProjects.map(untildify)];
  for (const project of allProjects) {
    const resolved = path.resolve(project);
    if (seen.has(resolved)) continue;
    seen.add(resolved);
    items.push(...scanProject(resolved));
  }

  return items;
}

module.exports = {
  GLOBAL_SCOPE,
  claudeHome,
  claudeJsonPath,
  discoverProjects,
  scanGlobal,
  scanProject,
  scanAll,
};
